### PR TITLE
Feature: add TopicTrait to filter repositories by topic

### PR DESCRIPTION
**Pull request by iwilltry42**
# Description

 This PR adds a new `TopicsTrait` which allows users to filter repositories by GitHub `topic`s (comma-separated list).

### Where this can be useful
In our organization, we're having multiple repositories where multiple teams have access to. If we filter by team-slug (which works really well :+1:), we'll end up with multiple pipeline jobs for the exact same repo, but in different organization/team folders in Jenkins. Then, webhooks trigger multiple builds of the same thing at once, which is not really optimal. With topics, one can filter by another dimension and e.g. create jobs only for repositories, which belong to `team: myteam` and have the `topic: managed-by-myteam`.

### Notes

- This is not a new list filter (i.e. not list-by-topic), but rather an existing list filter (i.e. filter-by-topic). That is filtering on Jenkins side, not on GitHub side.
- The Makefile is not really part of this PR, but I still think it could be a nice to have for people (like me) that are not too familiar with the Java and Maven ecosystem

### Possible Enhancements/Changes

- regex
- do search-by-topic, if no other search filter was set (e.g. not filtering by name or team-slug)
- always do search-by-topic instead of filter-by-topic and match repo lists

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate. --> None
- [x] Change is code complete and matches issue description
- [x] Automated tests have been added to exercise the changes
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below. --> None

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

Notifying @bitwiseman and @timja here, only because I saw them being tagged on another PR :grimacing:

### CODE REVIEW

**Comment by iwilltry42** [comment]
Hi @bitwiseman is there anything I can do to get this PR through? :)
Or did I just tag the wrong person?

**Comment by iwilltry42** [comment]
Somehow, the topics input is empty after Jenkins restart (e.g. after upgrading plugins).. can anyone give me a hint how this can be persisted?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java [hunk 1]
```diff
@@ -69,6 +76,21 @@ public String getTeamSlug() {
         return teamSlug;
     }
 
+    /**
+     * Sets the topics which the repositories must have.
+     */
+    void setTopics(ArrayList<String> topics) {
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java:82]: might fix your issue
```suggestion
    public void setTopics(ArrayList<String> topics) {
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java [hunk 2]
```diff
@@ -69,6 +76,21 @@ public String getTeamSlug() {
...
+        this.topics = topics;
+    }
+
+    /**
+     * Gets the topics which the repositories must have.
+     * @return topics
+     */
+    public ArrayList<String> getTopics() {
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java:90]: should the interface not the implementation
```suggestion
    public List<String> getTopics() {
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java (hunk 1 above)
Comment (iwilltry42) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java:82]: I just tested it and unfortunately, this change didn't help.
I probably don't understand enough on how the interaction between front- and backend works here... However, one can't even see the topics when just hitting `save` and then going back to `Configure`. The changes take effect though :thinking:

File: src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+  <f:entry title="${%Topics}" field="topicList">
```
Comment (timja) [inline src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly:4]: this is the data mapping issue
```suggestion
  <f:entry title="${%Topics}" field="topics">
```

Comment (iwilltry42) [reply src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly:4]: This makes sense, but then, `topicList` is a `String` while `topics` is an `ArrayList`.
In the `TopicsTrait` constructor, I then parse the ` String topicList` into the `ArrayList<String> topics` :thinking:

Comment (timja) [reply src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly:4]: ah i see

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java (hunk 1 above)
Comment (iwilltry42) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorContext.java]: It's fixed in commit 81bde9b

File: src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly (hunk 3 above)
Comment (iwilltry42) [reply src/main/resources/org/jenkinsci/plugins/github_branch_source/TopicsTrait/config.jelly:4]: Fixed in commit 81bde9b :+1:

**Comment by iwilltry42** [comment]
Hi there :wave: Is there anything I can do to improve this PR?
It's super cumbersome to always having to install a custom version here when having everything else defined via CasC :grimacing:

**Comment by timja** [comment]
I’m not a maintainer, @bitwiseman could you take a look?

Why is it cumbersome? How are you installing your plugins?

**Comment by iwilltry42** [comment]
> I’m not a maintainer, @bitwiseman could you take a look?
Oh, sorry, I just saw you being mentioned in some other PRs, that's why I tagged you :grimacing: 

> Why is it cumbersome? How are you installing your plugins?
I'm using the official Helm Chart and to enforce CasC, the plugins will be overwritten upon re-deployment. Then I have to go there and manually upload the custom .hpi file again to get this feature. Probably there are way easier solutions though :thinking:

**Comment by timja** [comment]
You can install it from the incrementals repo:

https://github.com/jenkinsci/docker#plugin-version-format

https://github.com/hmcts/cnp-jenkins-docker/blob/master/plugins.txt#L57

You need a green build for an incremental (basically a pre-release build) to be published

**Comment by iwilltry42** [comment]
Oh cool, I didn't know about the incrementals repo.
I'm actually not sure, why the build fails, but according to the logs it doesn't seem to be related to my changes: `[2020-06-12T01:02:11.487Z] EC2 (aws) - High memory ubuntu 18.04  (i-0056b6da636016e24) was marked offline: Connection was broken: java.io.EOFException`

**Comment by timja** [comment]
I've re-triggered it, you can do it by closing the PR, waiting a few seconds and then re-opening it

**Comment by iwilltry42** [comment]
Hmm.. build is green now, but it seems like it didn't trigger a deployment to the incrementals repo :thinking:

**Comment by timja** [comment]
you'll need to merge with master probably, it only deploys if its up-to date

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 4]
```diff
@@ -991,15 +998,20 @@ public void visitSources(SCMSourceObserver observer) throws IOException, Interru
...
 
                         if (repo.isArchived() && gitHubSCMNavigatorContext.isExcludeArchivedRepositories()) {
+                            // exclude archived repositories
                             witness.record(repo.getName(), false);
                             listener.getLogger()
                                     .println(GitHubConsoleNote.create(System.currentTimeMillis(), String.format(
                                             "Skipping repository %s because it is archived", repo.getName())));
-
-                        } else if (request.process(repo.getName(), sourceFactory, null, witness)) {
+                        } else if (!gitHubSCMNavigatorContext.getTopics().isEmpty() && !repo.listTopics().containsAll(gitHubSCMNavigatorContext.getTopics())) {
```
[the remark this answers is not in the document]
Comment (iwilltry42) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:1006]: Yep, this was also super confusing to me in the beginning :grimacing: 
Nothing to apologize though :)

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,9 @@
+.PHONY: test clean-install all
+
+all: test clean-install
+
+test:
+	docker run -it --rm --name my-maven-project -v "$(shell pwd)":/usr/src/mymaven -w /usr/src/mymaven maven:3-jdk-8 mvn test
+
+clean-install:
+	docker run -it --rm --name my-maven-project -v "$(shell pwd)":/usr/src/mymaven -w /usr/src/mymaven maven:3-jdk-8 mvn clean install
```
[the remark this answers is not in the document]
Comment (iwilltry42) [reply Makefile:9]: True, going to remove it :+1:

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java [hunk 6]
```diff
@@ -132,6 +132,19 @@ public void fetchOneRepo_BelongingToAuthenticatedUser() throws Exception {
         assertThat(projectNames, containsInAnyOrder("yolo-archived"));
     }
 
+    @Test
+    public void fetchRepos_BelongingToAuthenticatedUser_FilteredByTopic() throws Exception {
```
[the remark this answers is not in the document]
Comment (iwilltry42) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java:136]: done :heavy_check_mark:

**Comment by awsiv** [comment]
This would be very helpful.. the `team-slug` is kinda unusable because of this: https://issues.jenkins-ci.org/browse/JENKINS-62722

Even with that, I would still prefer this because of the flexibility it provides

**Comment by iwilltry42** [comment]
Good Morning @bitwiseman , are you OK with the provided tests or should I add some more? :+1:

**Pull request merged**